### PR TITLE
Promote edit

### PR DESCRIPTION
**Pull request by ericgiovannini**
Adds a new `promote` edit, which allows the user to specify that a definition that normally would be placed in the set of term-level definitions in the resulting Coq file should instead be "promoted" to the set of type-level definitions.

### CODE REVIEW

**Comment by lastland** [comment]
Does this change the way code is currently translated? The CI seems to suggest so, as the convenience copy of `base`, `containers`, `ghc`, and `transformers` seems to be different from what is generated with this patch. Can you try to rebuild the convenience copy? You should be able to do that by running `make clean && make` in their corresponding directories under `examples` (some of these may need you to rebuild `base-thy` as well, I don't remember exactly.) Let me know if you have any questions.

**Comment by lastland** [comment]
Oh I forgot, you can also simply run `boot.sh` from `examples` directory.

**Comment by ericgiovannini** [comment]
I tried running `boot.sh`, but it failed with the error ``can't find file: base/GHC/Base.hs``. So I then tried to build the base library by running `make -C base` from the root, but this immediately failed with `make: *** [all] Error 2`. Any ideas as to why this could be happening?

**Comment by lastland** [comment]
I'm not sure why `boot.sh` goes wrong, but can you try going to `examples/base-src` and run `make clean && make` from there? What `base` contains is the generated Coq code translated from the Haskell code in `examples/base-src`.

**Comment by ericgiovannini** [comment]
Just tried this, and I get a similar error about being unable to find `base/GHC/Base.hs`. Here are the last few lines that were displayed:

    ....
    stack exec hs-to-coq -- -e module-edits/GHC/Base/edits \
	              \
	             --midamble module-edits/GHC/Base/midamble.v \
		     -e edits -N -i gen-files -I gen-files -i base -I base/include --iface-dir ../../base --dependency-dir deps --ghc -this-unit-id --ghc base -o ../../base \
		     base/GHC/Base.hs
    hs-to-coq: panic! (the 'impossible' happened)
      (GHC version 8.4.3 for x86_64-apple-darwin):
	    can't find file: base/GHC/Base.hs

**Comment by lastland** [comment]
Uh sorry, I forget. The Haskell files are in submodules, you will need to run
```shell
git submodule update --init --recursive
```
The command will download all the submodules (some of them are from ghc).

**Comment by ericgiovannini** [comment]
Thanks, after running that, I was able to use `make clean && make` to rebuild the convenience copies for `base` (in `examples/base-src`), `containers`, and `transfomers`. But for `ghc`, the latter command failed with 

    make[2]: *** No rule to make target `../../containers/theories/MapProofs.vo', needed by `IntMap.vo'.  Stop.

**Comment by lastland** [comment]
It seems like you also need to compile `examples/containers/theories` to continue. The directory contains our theorems and proofs for the `containers` library, and it seems that our translation of `ghc` depends on it.

**Comment by ericgiovannini** [comment]
That seems to have fixed the error, but now a new error occurs:

    File "./CoreFVs.v", line 132, characters 58-81:
    Error: The reference bndrRuleAndUnfoldingFVs was not found in the current
    environment.

However, I repeated the above steps on a fresh clone of the repository and this error did not occur, so I am assuming it has to do with a change I made.

**Comment by lastland** [comment]
I do not recall seeing this error message in the past, so it is possible to be introduced by a change you made. I can take a look at it later (I was going to say tomorrow, but I should probably say during the daytime).

**Comment by ericgiovannini** [comment]
I've added two test cases, one (called Promote) that doesn't require recursive promotes, and the other (called Promote2) that does require them.

**Comment by lastland** [comment]
Thanks @ericgiovannini! It seems that these tests are a bit weak though, as they won't tell the promote edit from a no-op. I have added two `midamble.v` files to enhance the tests. `midamble.v` is inserted after type declarations but before value declarations, so it can be used to test if certain declarations have already appeared at this point.

File: doc/source/edits.rst [hunk 1]
```diff
@@ -849,6 +849,28 @@ Examples:
...
+
+.. index::
+   single: promote, edit
+
+Format:
+  | **promote** *qualified_name* ...
+
+Effect:
+  ``hs-to-coq`` divides Haskell definitions into two kinds: type-level and term-level.
+  The ``promote`` edit "prmotes" a term-level definition to the type level.
```
Comment (lastland) [inline doc/source/edits.rst:864]: I'm thinking if this description here would be a little bit misleading. People may incorrectly interpret this edit as something that works with the data type promotion extension of Haskell (i.e. `DataKinds`).

I think one thing we have not made clear in our previous documentation is that, in a translated Coq file, type-level definitions would appear before term-level definitions. This edit (if I understand correctly) does not try to change anything about the original definition. Instead, it moves the definition to the region dedicated to type-level definitions. Maybe the description for this edit can try to explain a bit of this? What do you think?

Comment (ericgiovannini) [reply doc/source/edits.rst]: Sure, I have modified the description for this edit to clarify this. Let me know if this is an improvement.

**Comment by lastland** [comment]
This looks good. Thank you @ericgiovannini!

File: src/lib/HsToCoq/ConvertHaskell/Parameters/Edits.hs [hunk 2]
```diff
@@ -266,7 +266,8 @@ subtractEdits edits1 edits2 =
   , _replacedTypes                  = edits1^.replacedTypes
   , _collapsedLets                  = edits1^.collapsedLets
   , _inEdits                        = edits1^.inEdits
-  , _exceptInEdits                  = edits1^.exceptInEdits 
+  , _exceptInEdits                  = edits1^.exceptInEdits
+  , _promotions                     = edits1^.promotions
```
Comment (lastland) [inline src/lib/HsToCoq/ConvertHaskell/Parameters/Edits.hs:270]: @ericgiovannini I am merging this PR with `master`. It turns out that I need to add this to make `Promote` edit work with `ExceptIn` edit. Do you think what I did here (in `subtractEdits`) make sense?

Comment (ericgiovannini) [reply src/lib/HsToCoq/ConvertHaskell/Parameters/Edits.hs:270]: Yes, this makes sense. Promoting a definition applies to the entire module, so it shouldn't be possible to localize the promote edit to a particular definition or exclude it from applying to a definition. In other words, the `_promotions` field is not used when translating individual definitions. So for `subtractEdits`, which is used when determining the edits that will be applied when translating an individual definition, the value of `_promotions` in the resulting edits record shouldn't matter.

Comment (lastland) [reply src/lib/HsToCoq/ConvertHaskell/Parameters/Edits.hs:270]: Great! Thanks. Merging in now.

**Pull request merged**